backend/common/src/main/java/ai/verta/modeldb/common/futures/Future.java

```diff
@@ -0,0 +1,387 @@
+
+  static {
+    String internalFutureTracingEnabled = System.getenv("IFUTURE_TRACING_ENABLED");
+    DEEP_TRACING_ENABLED = Boolean.parseBoolean(internalFutureTracingEnabled);
+  }
+
+  private final Tracer futureTracer = GlobalOpenTelemetry.getTracer("futureTracer");
+  private final String formattedStack;
+  private final CompletionStage<T> stage;
+
+  // keep the calling OpenTelemetry context, so we can use it to wrap the actual invocation of the
+  // future's implementation
+  private final Context callingContext = Context.current();
+
+  private Future(CompletionStage<T> stage) {
+    if (!DEEP_TRACING_ENABLED) {
+      formattedStack = null;
+    } else {
+      StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
+      if (stackTrace.length > 10) {
+        stackTrace = Arrays.copyOf(stackTrace, 10);
+      }
+      this.formattedStack =
+          Arrays.stream(stackTrace)
+              .map(StackTraceElement::toString)
+              .collect(Collectors.joining("\nat "));
+    }
+    this.stage = stage;
+  }
+
+  // Convert a list of futures to a future of a list
+  @SuppressWarnings("unchecked")
+  public static <T> Future<List<T>> sequence(final List<Future<T>> futures) {
+    if (futures.isEmpty()) {
+      return Future.of(List.of());
+    }
+
+    final var promise = new CompletableFuture<List<T>>();
+    final var values = new ArrayList<T>(futures.size());
+    final CompletableFuture<T>[] futuresArray =
+        futures.stream()
+            .map(x -> x.toCompletionStage().toCompletableFuture())
+            .collect(Collectors.toList())
+            .toArray(new CompletableFuture[futures.size()]);
+
+    CompletableFuture.allOf(futuresArray)
+        .whenCompleteAsync(
+            (ignored, throwable) -> {
+              if (throwable != null) {
+                promise.completeExceptionally(throwable);
+                return;
+              }
+
+              try {
+                for (final var future : futuresArray) {
+                  values.add(future.get());
+                }
+                promise.complete(values);
+              } catch (Throwable t) {
+                if (t instanceof InterruptedException) {
+                  // Restore interrupted state...
+                  Thread.currentThread().interrupt();
+                }
+                promise.completeExceptionally(t);
+              }
+            },
+            futureExecutor);
+
+    return Future.from(promise);
+  }
+
+  static <R> Future<R> from(CompletionStage<R> other) {
+    Preconditions.checkNotNull(
+        futureExecutor, "A FutureExecutor is required to create a new Future.");
+    return new Future<R>(other);
+  }
+
+  /** @deprecated Use {@link #of(Object)} instead. */
+  @Deprecated
+  public static <R> Future<R> completedInternalFuture(R value) {
+    Preconditions.checkNotNull(
+        futureExecutor, "A FutureExecutor is required to create a new Future.");
+    return of(value);
+  }
```

Comment:
should we just remove given that these are now two classes?

Reply:
it's a fair question. While doing a test migration from InternalFuture in UAC, I found this handy to have to make the conversion a little easier. I don't feel really strongly about it, though.